api.c

```diff
@@ -0,0 +1,434 @@
+        'k', 400, 401, 402, 403,
+        // Alignment
+        500, 501, 502, 503, 504, 505, 506,
+        // String stuff
+        510, 511, 512, 513, 514, 515, 516, 517, 518, 519,
+        // Fonts
+        520, 521, 522, 523, 524, 525,
+        // Text size
+        530, 531, 532, 533, 534, 535, 536,
+        // Text/indicator position
+        540, 541, 542, 543, 544, 545, 546, 547, 548,
+        // Text outline width
+        560, 561, 562, 563, 564, 565, 566,
+        // Bar indicator stuff
+        700, 701, 703, 704, 705, 706, 707, 708, 709,
+        710, 711,
+
+        0
+    };
+    size_t size = sizeof(enabled_options)/sizeof(*enabled_options);
+    api_options = malloc(size * sizeof(struct option));
+
+    int n = 0;
+    int i = 0;
+
+    while (n < size) {
+        if (longopts[i].val == enabled_options[n]) {
+            api_options[n] = longopts[i];
+            n++;
+        }
+        i++;
+    }
+}
+
+void *listen_api(void* lopts) {
+    struct option *longopts = lopts;
+
+    if (access(api_fifo_path, F_OK) == 0) {
+        if (remove(api_fifo_path) != 0) {
+            fprintf(stderr, "Could not remove file \"%s\": %s\n",
+                    api_fifo_path, strerror(errno));
+            return NULL;
+        }
+    }
+    if (mkfifo(api_fifo_path, S_IRWXU) != 0) {
+        fprintf(stderr, "Could not create FIFO file \"%s\": %s\n",
+                api_fifo_path, strerror(errno));
+        return NULL;
+    }
+
+    FILE *fifo;
+    char c;
+    int i;
+    int times = 0;
+    char *text = malloc(50);
+    arg_data data;
+
+    DEBUG("Created fifo at \"%s\"\n", api_fifo_path);
+
+    init_api_options(longopts);
+
+#define open_fifo() {\
+    fifo = fopen(api_fifo_path, "rb");\
+    if (fifo == NULL) {\
+        fprintf(stderr, "Could not open FIFO file \"%s\": %s\n", \
+                api_fifo_path, strerror(errno));\
+        return NULL;\
+    }\
+}
+#define close_fifo() {\
+    if (fifo != NULL) fclose(fifo);\
+}
+    open_fifo()
+    bool reading = true;
+
+    while (1) {
+        strcpy(text, "command ");
+        i = strlen(text);
+        int n = i;
+
+        reading = true;
+        while (reading && (c = fgetc(fifo)) != EOF) {
+            if (sizeof(text) <= i) {
+                text = realloc(text, i * 2);
+            }
+            if (c == '\n') {
+                text[i] = ' ';
+                reading = false;
+            } else
+                text[i] = c;
+            i++;
+        }
+        text[i] = '\0';
+        if (c == EOF) {
+            close_fifo();
+            open_fifo();
+        }
+        if (n == i) continue;
+
+        DEBUG("API message received - %d\n", times);
+        DEBUG("Message: \"%s\"\n", text);
+        times++;
+        if (parse_args(&data, text) != 0) {
+            fprintf(stderr, "Could not parse message\n");
+            continue;
+        }
+
+        update_api(data.argc, data.argv);
+        init_colors_once();
+
+        memset(text, 0, i);
```

Comment:
Is this needed?

Reply:
`update_api` updates the i3lock options, but the colors are not redrawn until i3lock decides to, so `init_colors_once` it's necessary to do an instant update.

The `memset` allows to "free" the `text` variable without de-allocating or creating new pointers, but I have some doubts about it.

Reply:
`memset` sets the memory, doesn't free it. You don't need to clear the memory to reuse it, just write over it in the next loop, then free it in the end of the function.

Reply:
Sure, it's not needed.